### ISSUE TITLE
Wrong compiler picked from tweeted short URL in Safari

### ISSUE THREAD

**Issue by mattgodbolt** [mattgodbolt/compiler-explorer#477]
Thread: https://twitter.com/vedantsk/status/882574073997520897 

The link resolves to:

https://gcc.godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(fontScale:1.1,j:1,options:(compileOnChange:'0'),source:'%23include+%3Climits.h%3E%0A%0Aint+foo(int+x)+%7B%0A++return+INT_MAX%2B1%3B%0A%7D'),l:'5',n:'1',o:'C%2B%2B+source+%231',t:'0')),k:49.10145036874587,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:g7snapshot,filters:(b:'0',commentOnly:'0',directives:'0',intel:'0'),options:'-O2+-fsanitize%3Dundefined'),l:'5',n:'0',o:'%231+with+x86-64+gcc+7+(snapshot)',t:'0')),k:50.898549631254134,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4

which seems to encode the `g7snapshot` compiler, which isn't what the reporter saw. I believe this is the intended compiler (@regehr ?).

Maybe something in the way the mobile is handling loading, but...not really sure. This is a mobile thing (I think...can't repro on Chrome), so sadly low priority at the moment until we have a decent solution for this.

**Comment by vedantk** [comment]
This appears to happen on Desktop Safari as well (macOS 10.12.1, Safari version 10.0.1 (12602.2.14.0.1)). Clicking on https://godbolt.org/g/2spSWN opens up my last CE session. After clicking "More -> Reset UI", pasting the link into the URL navigator, and hitting enter I see the default "square" function.

This is not an issue when using Google Chrome (59.0.3071.115) on the same machine.

I wonder how the '2spSWN' bit from the link is resolved into the encoded string you've posted above. Perhaps there's some logic in that code which pulls in the last open CE session the browser ran?

**Comment by mattgodbolt** [comment]
Very interesting! Thanks for checking out on Safari!

So: the 2spSWN bit is passed to `goo.gl` to be resolved by google's link shortener. Then a redirect to the full URL is provided, so that giant URL in the initial bug report is what (in theory!) you should be redirected to. Can you try browsing to https://goo.gl/2spSWN and seeing if you see the same problem. I'd guess you would.

Then try copy pasting the whole giant URL and checking that too. I also guess that would be a problem. Then we're down to the code itself.

In theory the code around here: https://github.com/mattgodbolt/compiler-explorer/blob/master/static/main.js#L114 should take the encoded settings from the URL and use them to override the currently-set settings (the whole `_.extent(defaultConfig, config();`

Can you check the Javascript console and see if there's an error there?

Thanks so much for any help: I don't have any macs around to test with!

**Comment by vedantk** [comment]
Ah gotcha. In Chrome (the browser which displays the correct result), https://goo.gl/2spSWN resolves to something 'interesting': "#g:!((g:!((g:!...[snip]". In Safari, the URL navigator shows such an 'interesting' URL for a split-second, before resolving to the bare "https://gcc.godbolt.org" URL. This explains why the last CE session opens up :).

In the Safari error console, I see a few messages, the most pertinent likely being:
'''Blocked a frame with origin "https://gcc.godbolt.org/" from accessing a frame with origin "https://platform.twitter.com/". Protocols, domains, and ports must match."'''

The other errors are variants on:
'''TypeError: undefined is not an object (evaluating 'safari.self.tab.dispatchMessage')''' (onloadwff.js:73)

**Comment by mattgodbolt** [comment]
Hmm. I suspect the twitter error is a red herring. I bet that's the "Tweet" icon in the "Other" menu at the top right failing to load for some reason.

The other bit I bet is due to the "interesting" format you allude to. I use rison encoding to make the link a little less "escapey". But maybe that's upsetting Safari as it's a little non-standard (though supported in other browsers).

If you try and create a CE link of your own, does that work? My guess is that no, it doesn't work for you from either a short or long link.

What version of Safari are you running?

Thanks again for your patience and help in tracking this down...

**Comment by vedantk** [comment]
I just created https://godbolt.org/g/tTHHKf from Desktop Safari, closed the CE tab, opened CE again to make edits to the last-saved session, and then browsed to the shortened link. This works! Browsing to the 'tTHHKf' link opens up the correct, unedited CE session.

I considered that there may be something fishy encoded in the link https://godbolt.org/g/2spSWN expands to. Not so! You helpfully posted the expanded version: I'm able to click on this and get the correct CE session from John's blog post.

I'm running Safari 10.0.1, which ships with macOS 10.12.1. I originally noticed the issue on MobileSafari from iOS 10.3.2. Glad to help track things down :).

Maybe the next question is: what is different between https://godbolt.org/g/2spSWN (John's link) and https://godbolt.org/g/tTHHKf (the one I just made)? Why does one resolve to a bare URL while the other resolves to the correct one? Could it be a Safari bug?
